agent/tests/test_lib_aws_sdk_php.c

```diff
@@ -6,12 +6,284 @@
+  tlib_pass_if_not_null("Expression should evaluate.", expr);
+  nr_php_zval_free(&expr);
+  nr_php_zval_free(&array_arg);
+
+  /*
+   *Invalid case: The argument array is not an array.  The wrapper should see
+   * the null return value.
+   */
+  char* array_arg_not_array = "1";
+  array_arg = tlib_php_request_eval_expr(array_arg_not_array);
+  expr = nr_php_call(NULL, "two_param", first_arg, array_arg);
+  tlib_pass_if_not_null("Expression should evaluate.", expr);
+  nr_php_zval_free(&expr);
+  nr_php_zval_free(&array_arg);
+
+  nr_php_zval_free(&first_arg);
+  tlib_php_request_end();
+  tlib_php_engine_destroy();
+}
+
+static inline void test_message_param_queueurl_settings_expect_val(
+    nr_segment_message_params_t message_params,
+    nr_segment_cloud_attrs_t cloud_attrs,
+    char* cloud_region,
+    char* cloud_account_id,
+    char* destination_name) {
+  tlib_pass_if_str_equal("cloud_region should match.", cloud_attrs.cloud_region,
+                         cloud_region);
+  tlib_pass_if_str_equal("cloud_account_id should match.",
+                         cloud_attrs.cloud_account_id, cloud_account_id);
+  tlib_pass_if_str_equal("destination_name should match.",
+                         message_params.destination_name, destination_name);
+
+  nr_free(cloud_attrs.cloud_region);
+  nr_free(cloud_attrs.cloud_account_id);
+  nr_free(message_params.destination_name);
+}
+
+static inline void test_message_param_queueurl_settings_expect_null(
+    nr_segment_message_params_t message_params,
+    nr_segment_cloud_attrs_t cloud_attrs) {
+  tlib_pass_if_null("cloud_region should be null.", cloud_attrs.cloud_region);
+  tlib_pass_if_null("cloud_account_id should be null.",
+                    cloud_attrs.cloud_account_id);
+  tlib_pass_if_null("destination_name should be null.",
+                    message_params.destination_name);
+}
+
+static void test_nr_lib_aws_sdk_php_sqs_parse_queueurl() {
+  /*
+   * nr_lib_aws_sdk_php_sqs_parse_queueurl extracts either ALL cloud_region,
+   * cloud_account_id, and destination_name or none.
+   */
+  nr_segment_message_params_t message_params = {0};
+  nr_segment_cloud_attrs_t cloud_attrs = {0};
+
+  tlib_php_engine_create("");
+
+// clang-format off
+#define VALID_QUEUE_URL  "https://sqs.us-east-2.amazonaws.com/123456789012/SQS_QUEUE_NAME"
+#define INVALID_QUEUE_URL_1 "https://us-east-2.amazonaws.com/123456789012/SQS_QUEUE_NAME"
+#define INVALID_QUEUE_URL_2 "https://sqs.us-east-2.amazonaws.com/123456789012/"
+#define INVALID_QUEUE_URL_3 "https://sqs.us-east-2.amazonaws.com/SQS_QUEUE_NAME"
+#define INVALID_QUEUE_URL_4 "https://random.com"
+#define INVALID_QUEUE_URL_5 "https://sqs.us-east-2.amazonaws.com/123456789012"
+#define INVALID_QUEUE_URL_6 "https://sqs.us-east-2.amazonaws.com/"
+#define INVALID_QUEUE_URL_7 "https://sqs.us-east-2.amazonaws.com"
+#define INVALID_QUEUE_URL_8 "https://sqs.us-east-2.random.com/123456789012/SQS_QUEUE_NAME"
+  // clang-format on
+
```

Comment:
Can you add tests where the 2nd and 3rd args are `NULL` instead of pointing to the message and cloud structs?

Reply:
In [this case](https://github.com/newrelic/newrelic-php-agent/pull/1003/files#diff-5cbabeb350f68acfaa71e353ed2387f1d2aa71b7b03201d820a283c3810e3aa2R220), the third argument queue_name would not exist.

In [this case](https://github.com/newrelic/newrelic-php-agent/pull/1003/files#diff-5cbabeb350f68acfaa71e353ed2387f1d2aa71b7b03201d820a283c3810e3aa2R224), we wouldn't decode the 2nd argument (account id).

If [one of the three is NULL, all will be NULL](https://github.com/newrelic/newrelic-php-agent/pull/1003/files#diff-b5410abe33a951704c8600c7b0b214d3c431cee94dcb3089567ec0ec9f328fc1R175); otherwise, the queueUrl is malformed and cannot be trusted.

Reply:
I mean something like
```
nr_lib_aws_sdk_php_sqs_parse_queueurl(NULL, NULL, NULL);
```

Reply:
Please see: [0230262](https://github.com/newrelic/newrelic-php-agent/pull/1003/commits/023026234bf80e79ef7d5f555087d290c4758ab8)